### PR TITLE
[GSoC'24] Refactor DeckPicker to Single Menu Layout

### PR DESCRIPTION
**Pull request by SanjaySargam**
## Purpose / Description
This PR aims to 

1. Convert deckpicker to have single menu
2. Remove StudyOptionsFragment when collection is empty
3. Remove back button when fragmented

## Approach
Moved menu items from DeckPicker screen to StudyOptionsFragment

## How Has This Been Tested?
Emulator

https://github.com/ankidroid/Anki-Android/assets/65113071/78a26dbf-eb91-4bef-ac08-9b01eb3b6bc9

Chromebook
[Screen recording 2024-05-22 11.08.56.webm](https://github.com/ankidroid/Anki-Android/assets/65113071/4723287c-5c48-4b97-94cd-319699dbacd9)

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — CHANGES_REQUESTED** [review]
* Could you split out the functional commits (hiding the fragment on onboarding) to make this easier to review and understand

I would have expected that the combination of menus would mean:

* Remove the menu from the fragment
* Have the activity control the menu, and select different menus depending on the fragment state

Is there a reason you didn't pick this implementation?

**Comment by SanjaySargam** [comment]
> Is there a reason you didn't pick this implementation?

If I do so, then the context menu ( three dots ) will show in the middle of the screen. I just thought having the menu at the top-right of the screen would be better so that's why I combined menus in fragment

File: AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt [hunk 1]
```diff
@@ -353,13 +354,13 @@ fun DeckPicker.shouldFetchMedia(preferences: SharedPreferences): Boolean {
 }
 
 suspend fun monitorMediaSync(
-    deckPicker: DeckPicker
+    context: Context
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:357]: Not able to access from fragment ( StudyOptionsFragment )

[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:357]: I'm sorry @SanjaySargam but I don't understand your answer. What can't access from, and how is StudyOptionsFragment related to this code change?

[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:357]: Ohhh sorry my bad. I'll revert this change

**Comment by Arthur-Milchior** [comment]
If I understand correctly David request, that means you'd split your PR into multiple commit.
The first commit would just take care of empty collection and remove the second fragment
the second commit would remove the back button when fragmented.
The third would move the menu from one fragment to the other

This way, it becomes far easier for us to review each commit, and consider whether it actually does what it is supposed to do

**Comment by SanjaySargam** [comment]
Now understood very well

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml [hunk 2]
```diff
@@ -1,7 +1,8 @@
-<menu xmlns:android="http://schemas.android.com/apk/res/android"
+<menu xmlns:tools="http://schemas.android.com/tools"
+    xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:ankidroid="http://schemas.android.com/apk/res-auto" >
     <item
-        android:id="@+id/action_undo"
+        android:id="@+id/action_undo1"
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/res/menu/study_options_fragment.xml:5]: Please, use names that are more meaningful.
In this case `action_undo_study_options` and you can rename the other one `action_undo_deck_picker`

Why do you need to have a different ID, while the items belows seems to be fine sharing the id of deck_picker ?

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml [hunk 3]
```diff
@@ -46,4 +47,69 @@
         android:id="@+id/action_unbury"
         android:title="@string/unbury"
         android:visibility = "gone"/>
+    <group android:id="@+id/allItems" android:visible="false">
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: My understanding is that this is a copy/paste of deck_picker content, right ?

I don't like to have duplicated code. Because, if one day we update one of the menu, we may forget to update the other one. 

Instead, in `StudyOptionsFragment.kt`, after `toolbar!!.inflateMenu(R.menu.study_options_fragment)` you can add `toolbar!!.inflateMenu(R.menu.deck_picker)`, and you'll have the content of deck_picker.xml at the end of study_option_fragment

Comment (SanjaySargam) [reply AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: > My understanding is that this is a copy/paste of deck_picker content, right ?

Yes

But the issue which I'm facing rn is search icon is showing in two screens ( DeckPicker & StudyOptionsFragment ) even when I explicitly hide the search icon in the fragment. Also the export menu is already there in fragment so I'm trying to hide export menu which is there in deck_picker.xml but still visible

**Comment by Arthur-Milchior** [comment]
I've a small problem right now with your code when I execute it. 
If you start on a fresh install, just after AnkiDroid asks you to grant access to storage, you start with a split view, and then a second later, you get a single view.
On the one paw, that's not that bad, because it'll occur only the first time you open ankidroid. On the other paw, that's exactly the time where we want to have a single view because the collection is empty. And that would even be worse because that will be the first thing the user see when they open the app on tablet.

I think the issue is that you only know whether you need to display one or two views once the collection is loaded. And obviously, it can't be loaded until permissions are given.
This does not seems to be an issue with the play variant. Probably because there is need for permission.

So, we need to decide what to do when we don't have access to the collection. That is, what to put in the screen before the collection is loaded. Whether there is one or two view by default.
The simplest solution, in my mind, would be to use a preference, as they will be imediately accessible, to recall whether we used to have content last time the collection was opened. It's default value would be "empty". And it would be set to non-empty when any behavior occurs to add a deck.
I'd love opinions of other maintainers, because that seems to be a lot of code complexity for something that will be a rare problem

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: I think, Sanjay, that you'll have to learn to give far more context in your text.
"Right now" seems strange to me. I don't know what "right now" refers to the code in your PR? Whether it refers to origin/main.
In all cases, I don't really understand what you mean here. I can't find any way to get two search icons. And I don't see it in your video.

Similarly, "the export menu is already there in fragment". There are two fragments so it was not clear to me which one you referred to, and I had to take a look to see it's in both fragments. It would be easier for me if you stated "both" directly

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 4]
```diff
@@ -223,7 +231,7 @@ class StudyOptionsFragment : Fragment(), ChangeManager.Subscriber, Toolbar.OnMen
 
     override fun onMenuItemClick(item: MenuItem): Boolean {
         when (item.itemId) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:233]: We can probably start the function with
```kotlin
if ((activity as DeckPicker).onOptionsItemSelected(item)) {
  return true;
}
```
So that we reuse all the cases considered in DeckPicker without duplicating code.
As a bonus, that would allow to avoid making some methods public

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 5]
```diff
@@ -335,34 +403,47 @@ class StudyOptionsFragment : Fragment(), ChangeManager.Subscriber, Toolbar.OnMen
             }
             // Switch on rename / delete / export if tablet layout
             if (fragmented) {
+                menu.setGroupVisible(R.id.allItems, true)
                 menu.findItem(R.id.action_rename).isVisible = true
                 menu.findItem(R.id.action_delete).isVisible = true
                 menu.findItem(R.id.action_export).isVisible = true
+                (activity as DeckPicker).setupMediaSyncMenuItem(menu)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:410]: If we use `(activity as DeckPicker)` that much, I suggest we introduce `private val deckPicker = activity as DeckPicker` and we use it everywhere. That'll be more readable I expect

File: AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:357]: Actually, you can access the DeckPicker from StudyOptionsFragment.kt. You already do it many time. It's just `activity as DeckPicker`

So you may not even need to change this function.

I agree with David however. We can just request it to be an Activity. We never actually need it to specifically be a DeckPicker.

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml (hunk 3 above)
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: I don't like that "allItems" does not contains all items from the menu. Is there a better way of describing what it contains ?

File: AnkiDroid/src/main/res/menu/deck_picker.xml [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
-    tools:context=".DeckPicker">
+    >
+    <item
+        android:id="@+id/deck_picker_action_filter"
+        android:icon="@drawable/ic_search_white"
+        android:title="@string/search_decks"
+        android:visible="false"
+        ankidroid:actionViewClass="androidx.appcompat.widget.SearchView"
+        ankidroid:showAsAction="always|collapseActionView"/>
     <group android:id="@+id/allItems">
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/res/menu/deck_picker.xml:12]: If we were to move the first element, I'd like the group to be renamed, as it's not "all item" anymore.
Also, I'd appreciate if you could explain in the commit message why you moved this item.
Generally, it'd be nice if you could explain the reason of any non-straightforward change in your commit message. And this is really not a change for which I can guess the meaning.

In particular, it seems that https://github.com/ankidroid/Anki-Android/blob/c2bb60dacabbbc7ea0ffab7b913b9d98b8e5d7b2/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt#L975 won't do any anymore to the filter, which seems quite a strange change to me

**Review by Arthur-Milchior — COMMENTED** [review]
There are still quite some improvements to do. I admit I really need more information and more context in order to really understand what you tried, what failed, and most importantly, why you did the change you did.

Please don't hesitate to message me on discord if you have questions about my remark. 

I still believe we can avoid that much code duplication.

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml (hunk 3 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: Can we name it as "commonActions" ?

Comment (SanjaySargam) [reply AnkiDroid/src/main/res/menu/study_options_fragment.xml:50]: Can we rename it as "secondaryActions"?

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 7]
```diff
@@ -2072,7 +2087,10 @@ open class DeckPicker :
         Timber.i("Updating deck list UI")
         hideProgressBar()
         // Make sure the fragment is visible
-        if (fragmented) {
+        if (fragmented && collectionIsEmpty) {
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt]:
```kotlin
if (fragmented) {
      invalidateOptionsMenu()
      studyoptionsFrame?.visibility = if (collectionIsEmpty) View.GONE else View.VISIBLE
  }
``` 

I've done this way, by the considering the case where you have only one deck and you deleted that deck, then the fragment will remove. And when we perform "undo delete deck" , the fragment will show but the optionsMenu is not updating

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 8]
```diff
@@ -2073,7 +2073,8 @@ open class DeckPicker :
         hideProgressBar()
         // Make sure the fragment is visible
         if (fragmented) {
-            studyoptionsFrame!!.visibility = View.VISIBLE
+            invalidateOptionsMenu()
+            studyoptionsFrame?.visibility = if (collectionIsEmpty) View.GONE else View.VISIBLE
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2091]: I'm not a fan of this change from `!!` to `?.`. I'd be interested in knowing the reason you did it.
If for some reason the frame was null, I'd be happy to have ankidroid crashing and let us know. 
Instead, this would be ignored, even while it should not occur.

If I'm missing something, I'd be interested to see it explained in the commit message.


I'll admit this !! is ugly. If you want to avoid it, I'd suggest creating a first commit where you state that the frame is lateinit. You'll then be able to just assume the variable is non-null everywhere

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 9]
```diff
@@ -363,6 +363,8 @@ class StudyOptionsFragment : Fragment(), ChangeManager.Subscriber, Toolbar.OnMen
                 icon!!.isAutoMirrored = true
                 toolbar!!.navigationIcon = icon
                 toolbar!!.setNavigationOnClickListener { (activity as AnkiActivity).finish() }
+            } else {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:390]: I'd appreciate if your commit message could contains slightly more details, so that I can understand what it does without having to read the code.

In this case, it would indicates that the button had no effect, and should not have been here in the first place.

Unless I'm wrong, this was a mistake in AnkiDroid, that was not related to the current PR.
I'd be fine putting it as a separate PR that could be merged very quickly.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 10]
```diff
@@ -970,17 +980,21 @@ open class DeckPicker :
...
         optionsMenuState?.run {
-            menu.findItem(R.id.deck_picker_action_filter).isVisible = searchIcon
             updateUndoLabelFromState(menu.findItem(R.id.action_undo), undoLabel)
             updateSyncIconFromState(menu.findItem(R.id.action_sync), this)
             menu.findItem(R.id.action_scoped_storage_migrate).isVisible = shouldShowStartMigrationButton
             setupMigrationProgressMenuItem(menu, mediaMigrationState)
         }
     }
 
+    private fun updateSearchFromState(menu: Menu) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:992]: Maybe, rename to `updateSearchVisibilityFromState` to be more specific.
Because, when I read the part of the code where the method is used, I was wondering whether it was updating the search result, the search field or anything else.

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:234]: Sorry, when re-reading my code snippet, I guess I should have added a comment too. Because, just reading it, it's not as clear as I first thought.

Something such as:
> // `item` was handled by the Deck Picker, no need to handle it here.

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 11]
```diff
@@ -81,6 +81,10 @@ class StudyOptionsFragment : Fragment(), ChangeManager.Subscriber, Toolbar.OnMen
     private lateinit var textTotal: TextView
     private var toolbar: Toolbar? = null
 
+    @VisibleForTesting
+    var createMenuJob: Job? = null
+    private lateinit var deckPicker: DeckPicker
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:86]: My bad, I gave you a wrong snippet in the previous remark.

What I was suggesting was 

```
   /** Current activity.*/
    private val deckPicker: DeckPicker
        get() = activity as DeckPicker
```

I'd consider putting it in its own commit.
The last commit is long enough, removing from the diff all lines where you use this value would simplify reviewing it

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml [hunk 12]
```diff
@@ -1,7 +1,7 @@
 <menu xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:ankidroid="http://schemas.android.com/apk/res-auto" >
     <item
-        android:id="@+id/action_undo"
+        android:id="@+id/action_undo_study_options"
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/res/menu/study_options_fragment.xml:4]: Similarly, to simplify reviewing the commit, I'd considering renaming the id in its own commit. So that I can concentrate on the important part of the review without having to see all those changes. 

You can do multiple renaming in a single commit.

File: AnkiDroid/src/main/res/menu/deck_picker.xml (hunk 6 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/res/menu/deck_picker.xml:12]: Why was this comment ignored? I admit I'm still confused by the name `allItems` that don't contain all items.

**Review by Arthur-Milchior — COMMENTED** [review]
Did you test your code also on small screen?
Because, we should not change anything on phone. And it seems your commit is doing some change. And actually, the orange icon and the empty line of code cause ankidroid to crash. 

Please don't hesitate to ask if you don't know how to debug it. Otherwise, I'd love if you could also have something that still works on android.

Appart from that, I mostly have small remarks concerning the style. It's still hard to review because last commit does a lot of things that are vaguely related. I really want to have some simple commits that do the renaming/introducing a value, that help read the code without any functional change.
And separately THE commit that do the actual functional change, that contains nothing but the functional change.

2024-06-09-044817_338x600_scrot
2024-06-09-044812_343x604_scrot

File: AnkiDroid/src/main/res/menu/deck_picker.xml (hunk 6 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/res/menu/deck_picker.xml:12]: Can we rename it as "secondaryActions"?

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt (hunk 9 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:390]: This was a mistake in AnkiDroid. I'll make separate PR

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 8 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2091]: > I'll admit this !! is ugly. If you want to avoid it, I'd suggest creating a first commit where you state that the frame is lateinit. You'll then be able to just assume the variable is non-null everywhere

we can't do that because on small screen `studyOptionsFrame` will be null

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt (hunk 11 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:86]: `java.lang.ClassCastException: com.ichi2.anki.StudyOptionsActivity cannot be cast to com.ichi2.anki.DeckPicker`

getting this error only on small screen

File: AnkiDroid/src/main/res/menu/study_options_fragment.xml (hunk 12 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/res/menu/study_options_fragment.xml:4]: It is not possibble to rename in separate commit . If I do so,  we will get an error because `action_undo` id contains in deck_picker menu and studyoptions menu

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 8 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt]: Then thanks for reverting this chanhge

**Review by Arthur-Milchior — APPROVED** [review]
All of my remarks are related to documentation/comment and can be handled later.
 I tested on smartphone and tablet simulator, and it seems to work as expected. So I'm fine approving it

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 13]
```diff
@@ -81,6 +81,13 @@ class StudyOptionsFragment : Fragment(), ChangeManager.Subscriber, Toolbar.OnMen
     private lateinit var textTotal: TextView
     private var toolbar: Toolbar? = null
 
+    @VisibleForTesting
+    var createMenuJob: Job? = null
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:85]: > How you intend to use it in test.
Really don't know, I just copied it from deckpicker . And I see `createMenuJob` is not used in any of the test.
Should I remove `@VisibleForTesting`?

**Comment by david-allison** [comment]
I think it's unusual that the menus for the fragmented view are the responsibility of the fragment on the right, as it adds complexity to the logic

I would expect the menu to be extracted out of the fragments, and extended to the full width of the activity (with the activity being responsible)

You would then have two menu files for the DeckPicker (fragmented/non), and one for the Options (non-fragmented), and but need to worry about conditionally applying text, or moving some of the processing to the fragment

Is there a technical reason why my hunch won't work out?

**Comment by SanjaySargam** [comment]
@david-allison Would it be acceptable to handle this as a TODO and merge as it is? I will work on this during GSOC period only

**Comment by david-allison** [comment]
It would, add a to-do and let's get this in

**Review by Arthur-Milchior — APPROVED** [review]
Fine by me. Only a very small nit. I'd be fine merging it as-is

**Review by david-allison — CHANGES_REQUESTED** [review]
Only a couple of string-related changes. The implementation looks good

Think about the ordering of this menu, and see if you can improve it

Screenshot 2024-07-05 at 16 00 06

**Review by david-allison — APPROVED** [review]
Cheers

**Comment by david-allison** [comment]
* https://github.com/ankidroid/Anki-Android/pull/16705
